pkg/network/netlink/consumer.go

```diff
@@ -0,0 +1,417 @@
+// It's also worth noting that in the event of an ENOBUF error, we'll re-create a new netlink socket,
+// and attach a BPF sampler to it, to lower the the read throughput and save CPU.
+func (c *Consumer) receive(output chan Event) {
+ReadLoop:
+	for {
+		c.pool.Reset()
+		msgs, err := c.socket.Receive()
+
+		if err != nil {
+			switch socketError(err) {
+			case errEOF:
+				// EOFs are usually indicative of normal program termination, so we simply exit
+				return
+			case errENOBUF:
+				// If we detect an ENOBUF during the initial Conntrack table dump it likely means
+				// the netlink socket recv buffer doesn't have enough capacity for the existing connections.
+				if !c.streaming {
+					log.Warnf("netlink: detected enobuf during conntrack table dump. consider raising rcvbuf capacity.")
+				}
+				atomic.AddInt64(&c.enobufs, 1)
+			}
+		}
+
+		throttlingErr := c.throttle(len(msgs))
+		if throttlingErr != nil {
+			return
+		}
+
+		// Messages with error codes are simply skipped
+		for _, m := range msgs {
+			if err := checkMessage(m); err != nil {
+				// TODO: Add some telemetry here
+				log.Debugf("netlink message error: %s", err)
+				continue ReadLoop
+			}
+		}
+
+		// Skip multi-part "done" messages
+		multiPartDone := len(msgs) > 0 && msgs[len(msgs)-1].Header.Type == netlink.Done
+		if multiPartDone {
+			msgs = msgs[:len(msgs)-1]
+		}
+
+		output <- c.eventFor(msgs)
+
+		// If we're doing a conntrack dump we terminate after reading the multi-part message
+		if multiPartDone && !c.streaming {
+			return
+		}
+	}
+}
+
+func (c *Consumer) eventFor(msgs []netlink.Message) Event {
+	return Event{
+		msgs:   msgs,
+		buffer: c.pool.inUse,
+		pool:   c.pool,
+	}
+}
+
+// throttle ensures that the read throughput from the socket stays below
+// the configured maxMessagePerSecond
+func (c *Consumer) throttle(numMessages int) error {
+	// We don't throttle the socket during initialization
+	// (when we dump the whole Conntrack table)
+	if !c.streaming {
+		return nil
+	}
+
+	c.breaker.Tick(numMessages)
+	if !c.breaker.IsOpen() {
+		return nil
+	}
+	atomic.AddInt64(&c.throttles, 1)
+
+	// Close current socket
+	// TODO: validate if we need to leave the group before creating a new socket
+	leaveErr := c.conn.LeaveGroup(netlinkCtNew)
+	if leaveErr != nil {
+		log.Errorf("netlink: error leaving group: %s", leaveErr)
+	}
+	c.socket.Close()
+
+	// Create new socket with the desired sampling rate
+	// We calculate the required sampling rate to reach the target maxMessagesPersecond
+	samplingRate := float64(maxMessagesPerSecond) * overshootFactor * c.samplingRate / float64(c.breaker.Rate())
+	err := c.initNetlinkSocket(samplingRate)
+	if err != nil {
+		log.Errorf("failed to re-create netlink socket. exiting conntrack: %s", err)
+		return err
+	}
+
+	// Reset circuit breaker
+	c.breaker.Reset()
+	// Re-subscribe netlinkCtNew messages
+	c.conn.JoinGroup(netlinkCtNew)
+
+	return nil
+}
+
+type bufferPool struct {
+	inUse *[]byte
```

Comment:
I'm a little confused as why the `inUse` field is here -- can't callers get a new buffer by calling `bufferPool#Get()`

Reply:
Yes, this is just a hack to not break the `netlink.Socket` interface, as there is no argument to the `Receive` method.
But I guess I could *add* a method `ReceiveInto([]byte)` and preserve the interface.